### PR TITLE
Fix missing stream synchronization in the threaded process group

### PR DESCRIPTION
**Pull request by albmalamd**
## Summary:

Threaded PG performs the data movement for all ranks on rank 0's thread, hence on rank 0's current stream. The current stream is thread-local, so a collective issued from a side stream is unordered w.r.t. rank 0's copies: rank 0 can read a rank's input before its copy-in has run, and a rank can read its output before rank 0 has written it. FSDP2 hits this on every collective because

FSDPCommContext.lazy_init gives each rank private all-gather and reduce-scatter streams. The host-side ordering in Collective.join was already correct, which is why the race went unnoticed for so long: it resolves favourably on NVIDIA CI and loses on ROCm, where it surfaces as check_sharded_parity gradient mismatches and diverging losses in the FSDP2 and replicate tests (ROCM-28104).

A real backend is stream-ordered w.r.t. the stream that issued the collective, and Collective.join now provides the same guarantee using events: every non-root rank records an event on the stream it issued from, rank 0 waits on those events before copying and records one afterwards, and the other ranks wait on that. The path is a no-op for CPU tensors. test_collectives_issued_from_side_stream covers the same ordering without involving FSDP.

## Reproduction:

The affected tests pass by luck on a fast enough GPU, so delay the producing side. In test_param_registration_after_forward, right before model(inp):

    if self.rank != 0:
        device_sleep(device_type.type, int(50 * get_cycles_per_ms(device_type.type)))

FSDP2's side streams wait on the rank's compute stream, so the sleep holds back the collective's inputs while rank 0 runs ahead. Without this change the test fails with 9/36 elements mismatched; with it, it passes. test_to_float64_after_init reproduces the same way with the sleep at the top of the training loop.

### CODE REVIEW

**Comment by jeffdaily** [comment]
Thanks for tracking this down, the diagnosis is right. I traced every `Collective.work()` implementation, including the cross-device `.to()` paths in `AllReduce` and `ReduceScatter` that launch on the source device's current stream, and the per-device event set covers all of them.

Rather than leave you a list of small items I pushed a follow-up commit (`83ad6f2`) directly to the branch. The synchronization semantics are unchanged; it is cleanup plus one test-robustness fix. Squash it, rework it, or tell me to drop it, whichever you prefer.

**The one that matters.** `get_cycles_per_ms()` is `lru_cache`d and was being called from inside the test body, which `MultiThreadedTestCase` runs concurrently on all four rank threads. Its calibration times ten `_sleep(1_000_000)` launches, so four concurrent calibrations serialize on the device: each thread measures roughly 4x the true elapsed time, computes a proportionally smaller cycles/ms, and whichever thread wins the cache stores that under-measured value for the rest of the process. The intended 200 ms delay shrinks to something closer to 50 ms. Every other in-tree caller of this helper is multi-process, so this is a new usage pattern. Rank 0 now calibrates behind a `dist.barrier()` and the others read the cache.

**Hot-path cost.** Threaded PG backs a large share of the DTensor and FSDP2 suites, and `join()` runs on every collective including the CPU-only ones where it can never produce an event. Each rank's device set is now computed once in `join()` and stashed in `self._devices`, and the leader unions those instead of running a second `pytree` walk over every rank's data.

**Readability.** The event containers are plain lists throughout instead of a mix of tuples and lists, and the reason rank 0 records no input event of its own (its copies already run on the stream that produced its data) is now stated in the Note rather than left implicit in the bare `rank > 0`.

**Test.** Dropped the explicit device index 0 to match the rest of the file, and switched to `all_gather_single` / `reduce_scatter_single` since the `*_into_tensor` / `*_tensor` spellings are deprecated and were emitting `FutureWarning`s from brand new code.

I considered pooling events to avoid creating and destroying `world_size + 1` of them per collective, and decided against it: the leader's post-work event could be re-recorded by its next collective before a slow rank has enqueued its wait, and if the leader switched streams in between, waiting on that recording would no longer imply the earlier work had completed. The churn is the safer trade.

On 4x gfx90a, verified the regression test still discriminates. With `multi_threaded_pg.py` reverted to its pre-fix state it fails with 8/32 mismatched elements, and passes with the fix:

```
python test/distributed/test_multi_threaded_pg.py -k test_collectives_issued_from_side_stream
```

Full file 23 passed; `test_fully_shard_init.py` (where `test_param_registration_after_forward` and `test_to_float64_after_init` live) 45 passed, 2 skipped; `test_functional_api.py` 9 errors, an identical set to the same run against an unmodified tree, all real-backend multi-GPU cases unrelated to this change.

```
python test/distributed/test_multi_threaded_pg.py
python test/distributed/_composable/fsdp/test_fully_shard_init.py
python test/distributed/test_functional_api.py
```

Two things left before this can land: it needs a `release notes:` or `topic: not user facing` label, and it is still marked draft.

_This comment was drafted with an AI assistant (Claude Code)._

**Review by jeffdaily — APPROVED** [review]

**Comment by jeffdaily** [comment]
@pytorchbot merge -i

Same periodic-rocm-mi200 failure on the merge base.
